Account for empty strings when splitting the host pattern

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Filter out empty strings when returning the list of hosts from `split_host_pattern()`.
Fixes #61964
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/inventory/manager.py`

samdoran committed:
Account for empty strings when splitting the host pattern

bcoca commented on lib/ansible/inventory/manager.py:
```diff
@@ -130,7 +130,7 @@ def split_host_pattern(pattern):
             )
 
-    return [p.strip() for p in patterns]
+    return [p.strip() for p in patterns if p.strip()]
```
to avoid dual stripping, change to 

```
res = []
for p in patterns:
     p = p.strip()
     if p:
        res.append(p)

return res
```

felixfontein reviewed (approved):
LGTM

s-hertel reviewed (approved):
shipit

bcoca reviewed (commented):
behaviour pre 2.9 (i checked 2.8-2.4) was always `'' == ','`,  which this restores

bcoca commented on test/integration/targets/inventory/runme.sh:
```diff
@@ -13,6 +13,14 @@ cleanup() {
+    echo "Inventory limit failed"
+    exit 1
+fi
+
```
you should also check that `--limit ''` and `--limit ','` don't change the list of hosts targetted

samdoran replied:
I added a whole slew of integration tests that also cover this scenario. Thanks for the suggestion.

samdoran committed:
Avoid double stripping

samdoran committed:
Improve tests

- add more unit test cases
- add specific integration test with more cases

bcoca commented on lib/ansible/inventory/manager.py:
```diff
@@ -129,8 +130,7 @@ def split_host_pattern(pattern):
             )
-
-    return [p.strip() for p in patterns]
+    return [p for p in patterns if p]
```
patterns from the regex won't be stripped?

also this traverses the list x2, not sure what is more expensive, strip x2 or list x2

samdoran replied:
Doh.

I went back to doing the "double strip" with a comprehension. With 100,000 hosts a regular `for` loop performed the same as doing the comprehension with the "double strip" on both Python 2 and 3.

```
patterns = [' host{} '.format(n) for n in range(1,100000)]

# Python 3
In [29]: %timeit split_host_pattern_loop(patterns)
1.06 s ± 66.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [30]: %timeit split_host_pattern_ds(patterns)
1.01 s ± 17.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


# Python 2
In [14]: %timeit split_host_pattern_loop(patterns)
1 loop, best of 3: 1.36 s per loop

In [18]: %timeit split_host_pattern_ds(patterns)
1 loop, best of 3: 1.37 s per loop
```

bcoca replied:
you REAAAALLY don't want a `for ` loop ....

samdoran committed:
Improve integration tests

samdoran committed:
Go back to using a comprehension

Testing shows no major downside to calling .strip() twice in a comprehnios vs. using a regular for loop and only calling .strip() once. Going with the comprehension for ease of maintenance and because comprehensions are optimized in CPython.
